### ISSUE TITLE
dll not found exception:mediapipe_c

### ISSUE THREAD

**Issue by yellowflash07** [homuler/MediapipeUnityPlugin#2]
hello, 
i have successfully built the .so file and placed it in SDK/Plugins, i have also downloaded the require tflite files and txt files and placed them under SDK/Models.
I am testing out the handtracking gpu on windows. when i press play i get:

DllNotFoundException: mediapipe_c
Mediapipe.CalculatorGraph..ctor (System.String configText) (at Assets/Mediapipe/SDK/Scripts/CalculatorGraph.cs:13)
DemoGraph.Initialize () (at Assets/Mediapipe/Examples/Scripts/DemoGraph.cs:22)
Director+<RunGraph>d__10.MoveNext () (at Assets/Mediapipe/Examples/Scripts/Director.cs:47)
UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) (at <3dc54541a2574ac7826a004a212a4332>:0)
UnityEngine.MonoBehaviour:StartCoroutine(IEnumerator)

and

DllNotFoundException: mediapipe_c
Director.OnEnable () (at Assets/Mediapipe/Examples/Scripts/Director.cs:19)

How to fix this?

**Comment by homuler** [comment]
DllNotFoundException is thrown in many cases so I'm not sure, but .so file might be broken.

Will you answer these questions?
1. Where did you build .so file and where is Unity run? (e.g. Ubuntu (WSL 2), Windows 10, etc...)
1. Please paste the command you run to build.
1. Can you run the official example?
https://google.github.io/mediapipe/solutions/hands

**Comment by yellowflash07** [comment]
1. I built the .so file with ubuntu wsl,windows 10 and unity is run on windows 10 as well
2. I used 
cd to navigate to mediapipeunity folder
bazel build -c opt --copt -DMESA_EGL_NO_X11_HEADERS --copt -DEGL_NO_X11 //mediapipe_api:mediapipe_c  
for building.
3. I was able to build following this tutorial : https://dev.classmethod.jp/articles/mediapipe-environment-setting-windows/

i am not able to build the official ones because of some errors in android sdk and ndk setup but that shouldn't ever.

**Comment by yellowflash07** [comment]
windows 10

**Comment by homuler** [comment]
OK. If you build the library on Linux, you have to run Unity on Linux too.

If you'd like to run Unity on Windows 10, you have to build the library on Windows 10 and in this case .dll file should be generated.
But unfortunately, I don't have time to test on Windows now, and some errors should be occurred.

**Comment by yellowflash07** [comment]
ok so how do i build a dll?

**Comment by yellowflash07** [comment]
hello, i changed my OS to ubuntu and i am still getting the error:

DllNotFoundException: mediapipe_c
Director.OnEnable () (at Assets/Mediapipe/Examples/Scripts/Director.cs:19)

DllNotFoundException: mediapipe_c
Mediapipe.CalculatorGraph..ctor (System.String configText) (at Assets/Mediapipe/SDK/Scripts/CalculatorGraph.cs:13)
DemoGraph.Initialize () (at Assets/Mediapipe/Examples/Scripts/DemoGraph.cs:22)
Director+<RunGraph>d__10.MoveNext () (at Assets/Mediapipe/Examples/Scripts/Director.cs:47)
UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) (at /home/bokken/buildslave/unity/build/Runtime/Export/Scripting/Coroutines.cs:17)
UnityEngine.MonoBehaviour:StartCoroutine(IEnumerator)
Director:SetWebCamDevice(Nullable`1) (at Assets/Mediapipe/Examples/Scripts/Director.cs:41)
DeviceSelectorController:OnValueChanged(Dropdown) (at Assets/Mediapipe/Examples/Scripts/DeviceSelectorController.cs:31)
DeviceSelectorController:ResetOptions(WebCamDevice[]) (at Assets/Mediapipe/Examples/Scripts/DeviceSelectorController.cs:25)
DeviceSelectorController:Start() (at Assets/Mediapipe/Examples/Scripts/DeviceSelectorController.cs:16)

DllNotFoundException: mediapipe_c
Director.OnDisable () (at Assets/Mediapipe/Examples/Scripts/Director.cs:29)

Unable to set resolution 1920 x 1080 for EasyCamera: EasyCamera (/dev/video0)
UnityEngine.WebCamTexture:Play()
WebCamScreenController:ResetScreen(Nullable`1) (at Assets/Mediapipe/Examples/Scripts/WebCamScreenController.cs:29)
Director:SetWebCamDevice(Nullable`1) (at Assets/Mediapipe/Examples/Scripts/Director.cs:35)
DeviceSelectorController:OnValueChanged(Dropdown) (at Assets/Mediapipe/Examples/Scripts/DeviceSelectorController.cs:31)
DeviceSelectorController:ResetOptions(WebCamDevice[]) (at Assets/Mediapipe/Examples/Scripts/DeviceSelectorController.cs:25)
DeviceSelectorController:Start() (at Assets/Mediapipe/Examples/Scripts/DeviceSelectorController.cs:16)


can you confirm handtracking gpu/cpu works on desktop?

**Comment by homuler** [comment]
Here is a sample video (running on GPU).
https://twitter.com/eulerdora/status/1302991294919180289

At latest commit (https://github.com/homuler/MediapipeUnityPlugin/commit/dd942dab54a27b597bb2f89f2d5b9ab5c6e8f82a), now it depends on protobuf and it became more difficult to build... but instead I added a Makefile to automate the build process (its code quality is not good but must be better than none).

By the way, can you run the official tutorial (handtracking) on your machine?
I also would like to know if you can run the [Hello World example](https://github.com/homuler/MediapipeUnityPlugin/tree/master/Assets/Mediapipe/Examples/HelloWorld) (please be sure to include **only** pass_through_calculator).

**Comment by yellowflash07** [comment]
Getting 49 errors as soon as I open the project.
Most of them are : 

The type or namespace name 'Google' could not be found in the global namespace (are you missing an assembly reference?)

Google.Protobuf.IMessage' in explicit interface declaration is not an interface

Assets\Mediapipe\SDK\Scripts\Formats\LocationData.cs(169,26): error CS0115: 'LocationData.Equals(object)': no suitable method found to override

Am I supposed to import some plugin? 

Also yes, the official handtracking is working for me.

**Comment by homuler** [comment]
> Am I supposed to import some plugin?

Yes. Please run `make` before starting UnityEditor or build protocol buffer by yourself.
cf. https://github.com/homuler/MediapipeUnityPlugin#build

**Comment by yellowflash07** [comment]
i get this when i run make

cd C && bazel build -c opt --copt -DMESA_EGL_NO_X11_HEADERS --copt -DEGL_NO_X11 //mediapipe_api:mediapipe_c && cd .. && mv -f C/bazel-bin/mediapipe_api/libmediapipe_c.so Assets/Mediapipe/SDK/Plugins
Starting local Bazel server and connecting to it...
INFO: Analyzed target //mediapipe_api:mediapipe_c (131 packages loaded, 5363 targets configured).
INFO: Found 1 target...
Target //mediapipe_api:mediapipe_c up-to-date:
  bazel-bin/mediapipe_api/libmediapipe_c.so
INFO: Elapsed time: 8.733s, Critical Path: 0.85s
INFO: 1 process: 1 linux-sandbox.
INFO: Build completed successfully, 2 total actions
cd Submodules/protobuf/csharp && ./buildall.sh && mv src/Google.Protobuf/bin/Release/net45/* ../../../Assets/Mediapipe/SDK/Plugins
/bin/sh: 1: cd: can't cd to Submodules/protobuf/csharp
make: *** [Makefile:20: Assets/Mediapipe/SDK/Plugins/Google.Protobuf.dll] Error 2


the last line says error 2. is that supposed to happen?

**Comment by homuler** [comment]
Please checkout submodules
https://git-scm.com/book/en/v2/Git-Tools-Submodules

**Comment by yellowflash07** [comment]
Ok, the errors are gone but now it's back to the old errors

DllNotFoundException: mediapipe_c
Director.OnEnable () (at Assets/Mediapipe/Examples/Scripts/Director.cs:19)

DllNotFoundException: mediapipe_c
Mediapipe.CalculatorGraph..ctor (System.String configText) (at Assets/Mediapipe/SDK/Scripts/CalculatorGraph.cs:13)
DemoGraph.Initialize () (at Assets/Mediapipe/Examples/Scripts/DemoGraph.cs:22)
Director+<RunGraph>d__10.MoveNext () (at Assets/Mediapipe/Examples/Scripts/Director.cs:47)
UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) (at /home/bokken/buildslave/unity/build/Runtime/Export/Scripting/Coroutines.cs:17)
UnityEngine.MonoBehaviour:StartCoroutine(IEnumerator)
Director:SetWebCamDevice(Nullable`1) (at Assets/Mediapipe/Examples/Scripts/Director.cs:41)
DeviceSelectorController:OnValueChanged(Dropdown) (at Assets/Mediapipe/Examples/Scripts/DeviceSelectorController.cs:31)
DeviceSelectorController:ResetOptions(WebCamDevice[]) (at Assets/Mediapipe/Examples/Scripts/DeviceSelectorController.cs:25)
DeviceSelectorController:Start() (at Assets/Mediapipe/Examples/Scripts/DeviceSelectorController.cs:16)

DllNotFoundException: mediapipe_c
Director.OnDisable () (at Assets/Mediapipe/Examples/Scripts/Director.cs:29)


Also, when I toggle on "Load on start up" in libmediapipe_c.so import settings the console shows:

Plugins: Couldn't open Assets/Mediapipe/SDK/Plugins/libmediapipe_c.so, error: Assets/Mediapipe/SDK/Plugins/libmediapipe_c.so: undefined symbol: _ZN2cv8fastFreeEPv

**Comment by homuler** [comment]
It looks like that opencv is not statically linked.
Which is your OpenCV version?

**Comment by yellowflash07** [comment]
pkg-config --modversion opencv returns 3.4.12

**Comment by homuler** [comment]
OK. The installation path may be different from me.
Currently, this project assumes that OpenCV is installed under `/usr/lib/`. 
https://github.com/homuler/MediapipeUnityPlugin/blob/master/C/third_party/linux_opencv_path.diff#L17

Is your OpenCV installed under `/usr/lib/x86_64-linux-gnu/` ?

**Comment by yellowflash07** [comment]
Yes, OpenCV is installed under /usr/lib/x86_64-linux-gnu/ and all the files exist as well

**Comment by homuler** [comment]
Will you try to comment out this line?
https://github.com/homuler/MediapipeUnityPlugin/blob/master/C/WORKSPACE#L21

**Comment by yellowflash07** [comment]
Now im getting:

Plugins: Couldn't open Assets/Mediapipe/SDK/Plugins/libmediapipe_c.so, error: Assets/Mediapipe/SDK/Plugins/libmediapipe_c.so: undefined symbol: _ZN2cv23getPerspectiveTransformERKNS_11_InputArrayES2_
UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr)

**Comment by homuler** [comment]
Hmm... Does ` /usr/lib/x86_64-linux-gnu/libopencv_imgproc.so` exist?

**Comment by yellowflash07** [comment]
yes it does

**Comment by homuler** [comment]
I don't know why it fails, but will you try installing those packages again?
https://github.com/google/mediapipe/blob/master/third_party/opencv_linux.BUILD#L9

If it continues to fail, please paste the result of below command.
```sh
objdump -T /usr/lib/x86_64-linux-gnu/libopencv_imgproc.so | grep getPerspectiveTransform
```
```

**Comment by homuler** [comment]
And please also restart UnityEditor after building the library.

**Comment by yellowflash07** [comment]
the error is gone. the issue was with the opencv installation folder. well there was a usr/lib/86_64-linux-gnu but the files were actually installed in usr/local/lib. This issue is in ubuntu 20.04 only
 
"Ubuntu 20.04 may install the OpenCV and FFmpeg packages in /usr/local" - from https://google.github.io/mediapipe/getting_started/install.html

After changing the workspace paths to /usr/local the dll exception was gone!

There are two more errors actually. relating to the camera.

Error mapping buffer (4147200): Invalid argument
UnityEngine.WebCamTexture:Play()

Unable to queue capture buffers!
UnityEngine.WebCamTexture:Play()

My webcam is working properly

**Comment by homuler** [comment]
> There are two more errors actually. relating to the camera.

The resolution settings might be invalid (e.g your camera may not support FHD (30FPS)).
https://github.com/homuler/MediapipeUnityPlugin/blob/master/Assets/Mediapipe/Examples/Scripts/WebCamScreenController.cs#L26

Will you try changing those values (DefaultHeight, DefaultWidth, FPS) from inspector window?

**Comment by yellowflash07** [comment]
The camera is running after changing resolution. 

Now im getting log errors:

Failed to fetch an output packet, rendering the input image
UnityEngine.Debug:LogWarning(Object)
DefaultGraphOnGPU:RenderOutput(Texture2D, Color32[]) (at Assets/Mediapipe/Examples/Scripts/DefaultGraphOnGPU.cs:19)
<RunGraph>d__10:MoveNext() (at Assets/Mediapipe/Examples/Scripts/Director.cs:67)
UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr) (at /home/bokken/buildslave/unity/build/Modules/IMGUI/GUIUtility.cs:197)

**Comment by homuler** [comment]
When you run a scene, some log files are created under `Logs/Mediapipe` and if you are lucky, mediapipe outputs some useful logs to them.
Are there any warning or error logs?

**Comment by yellowflash07** [comment]
Do you mean MedipipeUnityPlugin/Logs? if not where can I find Logs/Mediapipe? 

MedipipeUnityPlugin/Logs has only one log called packages update.log
which has


=== Wed Sep  9 13:34:44 2020

Packages were changed.
Update Mode: updateDependencies

The following packages were updated:
  com.unity.2d.animation from version 3.2.1 to 3.2.2
  com.unity.2d.psdimporter from version 2.1.3 to 2.1.4
  com.unity.2d.spriteshape from version 3.0.10 to 3.0.12
  com.unity.ide.vscode from version 1.2.0 to 1.2.1
  com.unity.test-framework from version 1.1.13 to 1.1.16

**Comment by homuler** [comment]
I'm sorry, I thought I checked in `Logs/Mediapipe/.gitkeep`, but this file is ignored.
Will you pull the latest commit (https://github.com/homuler/MediapipeUnityPlugin/commit/2a383badf2502d897ad507241c3fc94dd0903956) and rerun the scene again?

**Comment by homuler** [comment]
I'm not sure but it is possibly a bug. Will you create another issue, please?
I'd like to investigate more closely.

**Comment by yellowflash07** [comment]
Yes sure thank you very much for resolving the previous errors. Closing this issue now.

**Comment by homuler** [comment]
Thank you too for reporting the issue.
I'll update the document.
